From cf3053cef715d5762017102eaf0dc57fd20cf298 Mon Sep 17 00:00:00 2001
From: Yan Markman <ymarkman@marvell.com>
Date: Tue, 11 Nov 2025 15:56:09 +0200
Subject: [PATCH 1/1] platform/marvell-prest: Nokia7215-armhf ko-modules init

Add mvGpioDrv.ko source and build.
Run-time install mvcpss.ko and mvGpioDrv.ko by the nokia-7215init.sh.
This run-time install is strictly needed to speedup and resolve timing
latency and async-collisions on TRIXIE.

Signed-off-by: Yan Markman <ymarkman@marvell.com>
---
 .../sonic-platform-nokia/7215/scripts/nokia-7215init.sh  | 9 +++++++++
 .../marvell-prestera/sonic-platform-nokia/debian/rules   | 5 +++++
 2 files changed, 14 insertions(+)

diff --git a/platform/marvell-prestera/sonic-platform-nokia/7215/scripts/nokia-7215init.sh b/platform/marvell-prestera/sonic-platform-nokia/7215/scripts/nokia-7215init.sh
index 0df213537..1e35d4670 100755
--- a/platform/marvell-prestera/sonic-platform-nokia/7215/scripts/nokia-7215init.sh
+++ b/platform/marvell-prestera/sonic-platform-nokia/7215/scripts/nokia-7215init.sh
@@ -17,6 +17,13 @@ load_kernel_drivers() {
     modprobe eeprom
 }
 
+load_kernel_drivers_late() {
+    modprobe mvcpss
+    # Override mvGpioDrv.ko built-in mrvllibsai.deb with correct one
+    GOOD="/usr/lib/modules/$(uname -r)/kernel/extra/mvGpioDrv.ko"
+    find /var/lib/docker/overlay2/*/diff/usr/lib/modules -type f -name mvGpioDrv.ko \
+         -exec sh -c 'cp -a "$0" "$1" 2>/dev/null || true' "$GOOD" {} \;}
+}
 
 nokia_7215_profile()
 {
@@ -58,6 +65,8 @@ echo eeprom 0x56 > /sys/bus/i2c/devices/i2c-0/new_device
 echo eeprom 0x50 > /sys/bus/i2c/devices/i2c-1/new_device
 echo eeprom 0x51 > /sys/bus/i2c/devices/i2c-1/new_device
 
+load_kernel_drivers_late
+
 # Enable optical SFP Tx
 i2cset -y -m 0x0f 0 0x41 0x5 0x00
 
diff --git a/platform/marvell-prestera/sonic-platform-nokia/debian/rules b/platform/marvell-prestera/sonic-platform-nokia/debian/rules
index 7dde1c706..99abdc8a9 100755
--- a/platform/marvell-prestera/sonic-platform-nokia/debian/rules
+++ b/platform/marvell-prestera/sonic-platform-nokia/debian/rules
@@ -64,6 +64,8 @@ override_dh_auto_build:
 			cd $(MOD_SRC_DIR)/$${mod}/$(MRVL_MODULE_DIR)/$(PRESTERA_MODULE_SRC); \
 			make clean; \
 			make modules -C $(KERNEL_SRC)/build M=`pwd` CONFIG_KM_MVPCI=y CONFIG_KM_MVDMA=y CONFIG_KM_MVINT=y || exit 1; \
+			cd $(MOD_SRC_DIR)/$${mod}/$(MRVL_MODULE_DIR)/mvGpioDrv; \
+			make modules -C $(KERNEL_SRC)/build M=`pwd` || exit 1; \
 		fi; \
 		cd $(MOD_SRC_DIR)/$${mod}; \
 		python3 setup.py build; \
@@ -95,6 +97,9 @@ override_dh_auto_install:
 			cp $(MOD_SRC_DIR)/$${mod}/$(UTILS_DIR)/* debian/$(PACKAGE_PRE_NAME)-$${mod}/usr/local/bin/; \
 		fi; \
 		cp $(MOD_SRC_DIR)/$${mod}/$(MRVL_MODULE_DIR)/$(PRESTERA_MODULE_SRC)/mvcpss.ko debian/$(PACKAGE_PRE_NAME)-$${mod}/$(KERNEL_SRC)/$(INSTALL_MOD_DIR); \
+		if [ $$mod = "7215" ] && [ $(CONFIGURED_ARCH) = "armhf" ]; then \
+			cp $(MOD_SRC_DIR)/$${mod}/$(MRVL_MODULE_DIR)/mvGpioDrv/mvGpioDrv.ko debian/$(PACKAGE_PRE_NAME)-$${mod}/$(KERNEL_SRC)/$(INSTALL_MOD_DIR); \
+		fi; \
 		if [ -d "$(MOD_SRC_DIR)/$${mod}/$(MODULE_DIR)" ] && ls "$(MOD_SRC_DIR)/$${mod}/$(MODULE_DIR)"/*.ko >/dev/null 2>&1; then \
 			cp $(MOD_SRC_DIR)/$${mod}/$(MODULE_DIR)/*.ko debian/$(PACKAGE_PRE_NAME)-$${mod}/$(KERNEL_SRC)/$(INSTALL_MOD_DIR); \
 		fi; \
-- 
2.34.1

